pandas/indexes/base.py

```diff
@@ -1923,14 +1924,38 @@ def difference(self, other):
 
         other, result_name = self._convert_can_do_setop(other)
 
-        theDiff = sorted(set(self) - set(other))
-        return Index(theDiff, name=result_name)
+        this = self._get_unique_index()
+
+        # In what follows, get_indexer doesn't treat NaN's specially,
+        # so it would break the existing behavior, e.g.:
+        # `Index([1, nan]).difference(Index[nan]) == Index([nan, 1])`.
+        # Thus, dropping NaN's from `other` is a hack for backward compat.
+        # We do it here instead of manipulating `the_diff` later.
+        # (Check for MultiIndex is to get around a .hasnans exception)
+        dropna = not isinstance(self, ABCMultiIndex) and \
+            not isinstance(other, ABCMultiIndex) and \
+            self.hasnans and other.hasnans
```

Comment:
instead of this check, this is more idiomatic

```
try:
    other = other._get_unique_index(dropna=self.hasnan and other.hasnan)
except NotImplementeError:
   pass
```